community: added support for Guardrails for Amazon Bedrock

harelix opened the pull request:
Added support for optionally supplying 'Guardrails for Amazon Bedrock' on both types of model invocations (batch/regular and streaming) and for all models supported by the Amazon Bedrock service.

@baskaryan  @hwchase17

```python 
llm = Bedrock(model_id="<model_id>", client=bedrock,
                  model_kwargs={},
                  guardrails={"id": " <guardrail_id>",
                              "version": "<guardrail_version>",
                               "trace": True}, callbacks=[BedrockAsyncCallbackHandler()])

class BedrockAsyncCallbackHandler(AsyncCallbackHandler):
    """Async callback handler that can be used to handle callbacks from langchain."""

    async def on_llm_error(
            self,
            error: BaseException,
            **kwargs: Any,
    ) -> Any:
        reason = kwargs.get("reason")
        if reason == "GUARDRAIL_INTERVENED":
           # kwargs contains additional trace information sent by 'Guardrails for Bedrock' service.
            print(f"""Guardrails: {kwargs}""")


# streaming 
llm = Bedrock(model_id="<model_id>", client=bedrock,
                  model_kwargs={},
                  streaming=True,
                  guardrails={"id": "<guardrail_id>",
                              "version": "<guardrail_version>"})
```

harelix committed:
feature: guardrails for bedrock

harelix committed:
feature: guardrails for bedrock

harelix committed:
feature: guardrails for bedrock

harelix committed:
Merge branch 'langchain-ai:master' into bedrock-guardrails

gilinachum commented on libs/community/langchain_community/llms/bedrock.py:
```diff
@@ -191,6 +191,11 @@ class BedrockBase(BaseModel, ABC):
         "cohere": "stop_sequences",
     }
 
+    guardrails: Optional[Mapping[str, str]] = {
```
I would add a string doc on what this property represents (as it's not trivial)

harelix replied:
Thank you for pointing that out, @gilinachum. Recognizing the complexity of the property, I've updated it with a detailed docstring for better clarity on its purpose and usage.

gilinachum commented:
LGTM. Suggested to add a Docstring for the new Guardrails property.

harelix committed:
feature: guardrails for bedrock

harelix committed:
feature: guardrails for bedrock

harelix reviewed (commented):
I've updated it with a detailed docstring for better clarity on its purpose and usage.

Tom-HA commented:
+1

guya-pecan commented:
+1

romga commented:
👁️

harelix committed:
Merge branch 'master' into bedrock-guardrails

harelix committed:
Merge branch 'langchain-ai:master' into bedrock-guardrails

harelix committed:
Merge branch 'langchain-ai:master' into bedrock-guardrails

harelix committed:
bedrock guardrails trace callbacks

harelix committed:
bedrock guardrails trace callbacks

harelix committed:
bedrock guardrails trace callbacks

harelix committed:
bedrock guardrails trace callbacks

harelix committed:
Merge branch 'langchain-ai:master' into bedrock-guardrails

harelix committed:
bedrock guardrails trace callbacks

harelix commented on libs/community/langchain_community/llms/bedrock.py:
```diff
@@ -191,6 +201,52 @@ class BedrockBase(BaseModel, ABC):
+    To enable tracing for guardrails, set the 'trace' key to True and pass a callback handler to the 'run_manager' parameter of the 'generate', '_call' methods.
+
+    Example:
+    llm = Bedrock(model_id="<model_id>", client=<bedrock_client>,
+                  model_kwargs={},
+                  guardrails={
+                        "id": "<guardrail_id>",
+                        "version": "<guardrail_version>",
+                        "trace": True},
+                callbacks=[BedrockAsyncCallbackHandler()])
+
+    [https://python.langchain.com/docs/modules/callbacks/] for more information on callback handlers.
+
+    class BedrockAsyncCallbackHandler(AsyncCallbackHandler):
+        async def on_llm_error(
+            self,
+            error: BaseException,
+            **kwargs: Any,
+        ) -> Any:
+            reason = kwargs.get("reason")
+            if reason == "GUARDRAIL_INTERVENED":
+                ...Logic to handle guardrail intervention...
+    """
```
Guardrails for Bedrock Intervention Callback Handler

hwchase17 commented:
@harelix  and @gilinachum i havent had time to test this, but am i correct in assuming that you both are from amazon? and i should feel confident just merging this in and trusting your testing/code review?

harelix committed:
bedrock guardrails trace callbacks

harelix committed:
Merge branch 'langchain-ai:master' into bedrock-guardrails

harelix committed:
Merge branch 'langchain-ai:master' into bedrock-guardrails

harelix committed:
add Bedrock Guardrails full tracing and documentation

harelix committed:
add Bedrock Guardrails full tracing and documentation

harelix committed:
Merge branch 'master' into bedrock-guardrails

harelix committed:
Merge branch 'langchain-ai:master' into bedrock-guardrails

harelix commented:
@hwchase17 true, we are both from Amazon, and we run our tests to verify that this is actually working and not introducing breaking changes.

harelix committed:
Merge branch 'master' into bedrock-guardrails

harelix committed:
Merge branch 'master' into bedrock-guardrails

harelix committed:
Merge branch 'master' into bedrock-guardrails

harelix committed:
Merge branch 'master' into bedrock-guardrails

harelix committed:
Merge branch 'master' into bedrock-guardrails

harelix committed:
Merge branch 'master' into bedrock-guardrails

harelix committed:
Merge branch 'master' into bedrock-guardrails

3coins reviewed (commented):
@harelix
Thanks for adding this. There are some minor comments, and naming suggestions. Are you planning to add any unit tests? 
Here is a good PR to review for unit tests.
https://github.com/langchain-ai/langchain/pull/12477

harelix commented on libs/community/langchain_community/llms/bedrock.py:
```diff
@@ -254,6 +312,43 @@ def _get_provider(self) -> str:
+            raise TypeError("Guardrails must be a dictionary with 'id' and 'version' keys.") from e
+
+
+    def _get_guardrails_canonical_form(self) -> Dict[str, Any]:
```
done, please review.

harelix committed:
Merge branch 'langchain-ai:master' into bedrock-guardrails

harelix committed:
post review: refactor functions names and signatures

harelix commented on libs/community/langchain_community/llms/bedrock.py:
```diff
@@ -284,8 +391,58 @@ def _prepare_input_and_invoke(
+        if body.get(GUARDRAILS_BODY_KEY) == "GUARDRAIL_INTERVENED":
+            return True
+
+        return False
```
done, please review

harelix commented on libs/community/langchain_community/llms/bedrock.py:
```diff
@@ -284,8 +391,58 @@ def _prepare_input_and_invoke(
+                    "signal" : True,
+                    "reason" : "GUARDRAIL_INTERVENED",
+                    "trace" : body.get(AMAZON_BEDROCK_TRACE_KEY)
+                }
```
done, please review

harelix commented on libs/community/langchain_community/llms/bedrock.py:
```diff
@@ -284,8 +391,58 @@ def _prepare_input_and_invoke(
+           body dict():
+
+        Returns:
+           str: The text that was generated by the model.
```
done, please review

harelix commented on libs/community/langchain_community/llms/bedrock.py:
```diff
@@ -284,8 +391,58 @@ def _prepare_input_and_invoke(
+
+
+
+    def _check_if_guardrails_intervened(self, body: dict) -> bool:
```
done, please review

harelix commented on libs/community/langchain_community/llms/bedrock.py:
```diff
@@ -284,8 +391,58 @@ def _prepare_input_and_invoke(
         return text
 
+
+    def _inspect_bedrock_services_signal(self, body: dict) -> dict:
```
This method was originally composed as a general-purpose signal inspection, so I have changed it to: `_get_bedrock_services_signal`

harelix commented:
> @harelix Thanks for adding this. There are some minor comments, and naming suggestions. Are you planning to add any unit tests? Here is a good PR to review for unit tests. #12477

I will include my integration tests in this PR

harelix commented:
> @hwchase17 true, we are both from Amazon, and we run our tests to verify that this is actually working and not introducing breaking changes.

@hwchase17 - I've addressed the feedback and tested the changes in my pull request, and It's ready to be merged. Would you mind taking a look and merging it when you have a moment? I'd really appreciate it!

harelix committed:
Merge branch 'langchain-ai:master' into bedrock-guardrails

harelix committed:
bedrock guardrails integration tests

3coins reviewed (approved):
@harelix 
Thanks for working on these changes. LGTM!

gilinachum reviewed (approved):
LGTM

baskaryan commented:
apologies for the delay! code looks good but looks like there's now some merge conflicts. happy to merge once those are resolved

harelix committed:
merged with Bedrock async methods #12477

harelix committed:
Merge branch with Bedrock async methods #12477 into bedrock-guardrails

harelix committed:
Merge branch 'langchain-ai:master' into bedrock-guardrails

harelix committed:
Merge branch 'langchain-ai:master' into bedrock-guardrails

harelix committed:
fix: merge conflict resolution

harelix committed:
fix: indentation on stream

harelix committed:
fix: removing duplicate

harelix committed:
fix: remove duplicate

baskaryan committed:
Merge branch 'master' into bedrock-guardrails

baskaryan committed:
fmt

harelix committed:
Merge branch 'master' into bedrock-guardrails

harelix committed:
fix: notebook bedrock client

harelix commented:
>  code looks good but looks like there's now some merge conflicts. happy to merge once those are resolved

@baskaryan;  resolved - thank you!

harelix committed:
Merge branch 'langchain-ai:master' into bedrock-guardrails

baskaryan committed:
Merge branch 'master' into bedrock-guardrails

baskaryan committed:
fmt

baskaryan committed:
fmt
